crates/wasi/src/preview2/ip_name_lookup.rs

```diff
@@ -5,35 +5,73 @@ use crate::preview2::poll::{subscribe, Pollable, Subscribe};
 use crate::preview2::{spawn_blocking, AbortOnDropJoinHandle, SocketError, WasiView};
 use anyhow::Result;
 use std::mem;
-use std::net::{Ipv6Addr, ToSocketAddrs};
+use std::net::{IpAddr, Ipv6Addr, ToSocketAddrs};
 use std::pin::Pin;
 use std::str::FromStr;
 use std::vec;
 use wasmtime::component::Resource;
 
-use super::network::{from_ipv4_addr, from_ipv6_addr};
+use super::WasiCtx;
 
 pub enum ResolveAddressStream {
-    Waiting(AbortOnDropJoinHandle<Result<Vec<IpAddress>, SocketError>>),
-    Done(Result<vec::IntoIter<IpAddress>, SocketError>),
+    Waiting(AbortOnDropJoinHandle<std::io::Result<Vec<IpAddr>>>),
+    Done(std::io::Result<vec::IntoIter<IpAddr>>),
 }
 
-#[async_trait::async_trait]
-impl<T: WasiView> Host for T {
+/// A trait for providing a `IpNameLookup` implementation.
+pub trait WasiNetworkView: Send {
+    /// Given a name, resolve to a list of IP addresses
     fn resolve_addresses(
```

Comment:
This might be best to model as an `async` function rather than returning an `AbortOnDropJoinHandle`?

Reply:
This is indeed very annoying... The value returned from `resolve_address` must be `Send + Sync` as it is used in `ResolveAddressStream` which impls `Subscribe` which requires the type implementing it to be `Send + Sync`.

The `#[async_trait]` macro puts a `Send` bound on the return type but not `Sync` and there is no way to instruct it to do so. We could move to using `Box<dyn Future>` but that means we loose the developer experience of `async_trait`. We could potentially move to using `impl Future` but that requires the latest version of Rust, and I'm not sure if that's ok.

Reply:
Hm that's true, and yeah I see how there's no great way to manage this right now.

I suppose this can stay for now, but I'm still worried that this will inevitably cause some pain down the road. The best alternative I can think of is to return a `oneshot::Receiver<T>` instead but that's still pretty cumbersome to work with and wouldn't work great here.

Reply:
We could take a page out of `WasiHttpView`'s playbook and have the function return a `wasmtime::Result<Resource<ResolveAddressStream>>` just like `WasiHttpView::send_request` returns a `wasmtime::Result<Resource<HostFutureIncomingResponse>>`. I don't really love this either, but at least it's consistent?

Reply:
Of all the options mentioned so far, the `Box<dyn Future>` route seems the least worst to me.

The fact that we currently use a quick&dirty implementation using spawn_blocking should be an implementation detail of `SystemNetwork`

Reply:
Returning a `Resource<ResolveAddressStream>` would be nice yeah but it seems like that would have all the same design questions around "how do you actually create one of those". 

Although as I say that we could perhaps get around `async trait` by having something like `fn from_future<T: Future + Send + Sync + 'static> (...)` on the `ResolveAddressStream` type. That way you could pass an `async` block which would have bounds generated correctly and additioanlly the handle from `spawn_blocking` could also be passed in? That way we might be able to avoid async traits altogether and rely entirely on that.

Reply:
Thinking about this more, returning `Resource<ResolveAddressStream>` would mean that the implementor would have to also implement `WasiCtx` or otherwise have multiple access to the `ResourceTable`. Right now we don't require that from the implementor of `WasiNetworkView`, and I'm not sure we want to.

Reply:
`ResolveAddressStream` itself could be returned though leaving the table management bits to the glue?